src/desktop/space.rs

```diff
@@ -0,0 +1,419 @@
+                    Some(*w)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<Rectangle<i32, Logical>>>()
+        {
+            slog::debug!(self.logger, "Removing window at: {:?}", old_window);
+            damage.push(old_window);
+        }
+
+        // lets iterate front to back and figure out, what new windows or unmoved windows we have
+        for window in self.windows.iter().rev() {
+            let geo = window_rect(window, &self.id);
+            let old_geo = state.last_state.get(&window.0.id).cloned();
+
+            // window was moved or resized
+            if old_geo.map(|old_geo| old_geo != geo).unwrap_or(false) {
+                // Add damage for the old position of the window
+                damage.push(old_geo.unwrap());
+            } /* else {
+                  // window stayed at its place
+                  // TODO: Only push surface damage
+                  //       But this would need to take subsurfaces into account and accumulate damage at least.
+                  //       Even better would be if damage would be ignored that is hidden by subsurfaces...
+              }*/
+            // Add damage for the new position (see TODO above for a better approach)
+            damage.push(geo);
+        }
+
+        // That is all completely new damage, which we need to store for subsequent renders
+        let new_damage = damage.clone();
+        // We now add old damage states, if we have an age value
+        if age > 0 && state.old_damage.len() >= age {
+            // We do not need older states anymore
+            state.old_damage.truncate(age);
+            damage.extend(state.old_damage.iter().flatten().copied());
+        } else {
+            // just damage everything, if we have no damage
+            damage = vec![output_geo];
+        }
+
+        // Optimize the damage for rendering
+        damage.retain(|rect| rect.overlaps(output_geo));
+        damage.retain(|rect| rect.size.h > 0 && rect.size.w > 0);
+        for rect in damage.clone().iter() {
+            // if this rect was already removed, because it was smaller as another one,
+            // there is no reason to evaluate this.
+            if damage.contains(rect) {
+                // remove every rectangle that is contained in this rectangle
+                damage.retain(|other| !rect.contains_rect(*other));
+            }
+        }
+
+        let output_transform: Transform = output.current_transform().into();
+        if let Err(err) = renderer.render(
+            output_transform
+                .transform_size(output_size)
+                .to_f64()
+                .to_physical(state.render_scale)
+                .to_i32_round(),
+            output_transform,
+            |renderer, frame| {
+                // First clear all damaged regions
+                for geo in &damage {
+                    slog::debug!(self.logger, "Clearing at {:?}", geo);
```

Comment:
Maybe put this as `trace!` instead? That may be extremely spammy

Reply:
Locally I have already changed that to trace, it was just super helpful debugging this.